src/modules/hunting/apiserver.py

```diff
@@ -0,0 +1,716 @@
+
+    def __init__(self, evidence):
+        Vulnerability.__init__(self, KubernetesCluster, name="Patched A Pod",
+                               category=AccessRisk)
+        self.evidence = evidence
+
+
+class DeleteAPod(Vulnerability, Event):
+    """ Deleting a pod allows an attacker to disturb applications on the cluster """
+
+    def __init__(self, evidence):
+        Vulnerability.__init__(self, KubernetesCluster, name="Deleted A Pod",
+                               category=AccessRisk)
+        self.evidence = evidence
+
+
+class ApiServerPassiveHunterFinished(Event):
+    def __init__(self, all_namespaces_names, service_account_token, host, port):
+        self.host = host
+        self.port = port
+        self.all_namespaces_names = all_namespaces_names
+        self.service_account_token = service_account_token
+
+    def __str__(self):
+        return str(self.service_account_token)
+
+
+# Passive Hunter
+@handler.subscribe(OpenPortEvent, predicate=lambda x: x.port == 443 or x.port == 6443)
+class AccessApiServerViaServiceAccountToken(Hunter):
+    """ API Server Hunter
+    Accessing the API server within a compromised pod might grant an attacker full control over the cluster
+    """
+
+    def __init__(self, event):
+        self.event = event
+        self.api_server_evidence = ''
+        self.service_account_token_evidence = ''
+        self.pod_list_under_default_namespace_evidence = ''
+        self.pod_list_under_all_namespaces_evidence = ''
+
+        self.all_namespaces_names_evidence = list()
+        self.all_roles_names_evidence = list()
+        self.roles_names_under_default_namespace_evidence = list()
+        self.all_cluster_roles_names_evidence = list()
+        self.namespaces_and_their_pod_names = list(dict())
+
+    def access_api_server(self):
+        logging.debug(self.event.host)
+        logging.debug('Passive Hunter is attempting to access the API server using the pod\'s service account token')
+        try:
+            res = requests.get("https://{host}:{port}/api".format(host=self.event.host, port=self.event.port),
+                               headers={'Authorization': 'Bearer ' + self.service_account_token_evidence}, verify=False)
+            self.api_server_evidence = res.content
+            return res.status_code == 200 and res.content != ''
+        except requests.exceptions.ConnectionError:
+            return False
+
+    def get_service_account_token(self):
+        logging.debug(self.event.host)
+        logging.debug('Passive Hunter is attempting to access pod\'s service account token')
+        try:
+            with open('/var/run/secrets/kubernetes.io/serviceaccount/token', 'r') as token:
+                data = token.read()
+                self.service_account_token_evidence = data
+                return True
+        except IOError:  # Couldn't read file
+            return False
+
+    # 2 Pods Methods:
+    def get_pods_list_under_default_namespace(self):
```

Comment:
In fact there are a lot of cases here where you do something very like this:

```
try: 
    # build a request
    res = <response from request>
    # do something with response
    return res.content if res.status_code == 200 else False
except (requests.exceptions.ConnectionError, KeyError):
    return False
```

Can you make this cleaner without making it harder to read? Have a look at the kubelet hunter code for inspiration. You could have a class that looks after the basic logic, with the part where you build the request and the part where you handle the response defined for subclasses... 

This may be less of an issue if you can get rid of the quote substitution and the ascii encoding, as there will be less repetitious code.